1-Draft/RFC####-Improve-generation-of-argument-string-for-executables.md

```diff
@@ -0,0 +1,85 @@
+---
+
+# Improve generation of argument-string for executables
+
+When PowerShell calls an external executable, the
+[NativeCommandParameterBinder](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/NativeCommandParameterBinder.cs)
+takes all parameters of the command and generates the argument-string (also called "CommandLine"), which is sent to the executable (as [ProcessStartInfo.Arguments](https://msdn.microsoft.com/en-us/library/system.diagnostics.processstartinfo.arguments.aspx)).
+It seems like the NativeCommandParameterBinder tries to generate an argument-string from the given parameters, such that argv[] of the called executable is set accordingly. This works well, if an argument contains spaces. In this case the NativeCommandParameterBinder adds quotes around the argument, so it appears in `argv[]` as one argument (and is not split into parts).
+However, if the argument itself contains quotes, these are not escaped. Therefore, the corresponding element in `argv[]` has no quotes and depending on the actual string, the argument might be split into multiple `argv[]` elements. It can even occur, that the following arguments are not handled correctly.
+
+This RFC suggests making the NativeCommandParameterBinder compatible to
+[the typical CommandLine escaping rules](https://msdn.microsoft.com/en-us/library/17w5ykft.aspx).
+Additionally, it suggests adding a verbose-symbol to add a custom formatted string to the CommandLine (for executables, that don't follow the typical escaping rules.)
+
+This is basically a bugfix for https://github.com/PowerShell/PowerShell/issues/1995 and https://github.com/PowerShell/PowerShell/issues/3049 and on stackoverflow (["This seems like a bug to me. If I am passing the correct escaped strings to PowerShell, then PowerShell should take care of whatever escaping may be necessary for however it invokes the command."](http://stackoverflow.com/questions/6714165/powershell-stripping-double-quotes-from-command-line-arguments)
+and ["... and I think this is a bug Powershell doesn't escape any double quotes that appear inside the arguments."](http://stackoverflow.com/a/21334121/2770331)).
+However, as this is a very longstanding bug, there are workarounds for some cases. These will no longer work if the proposed changes are applied, therefore this document suggests to add a preference variable to get the old behavior back if somebody needs it.
+
+## Motivation
+
+    As a PowerShell user who wants to call external executables,
+    I can pass any arguments to these executables,
+    so that they reach the `argv[]` array unchanged.
+
+## Specification
+
+### Quoting
+
+The decision if an argument needs to be quoted, will be simplified: Any Argument that contains `"`, `'` or a character that matches `char.IsWhiteSpace` will be quoted. To quote an argument (compatible to [MSVC rules](https://msdn.microsoft.com/en-us/library/17w5ykft.aspx) and `CommandLineToArgvW`):
```

Comment:
It seem to me that  `'`-containing tokens don't need quoting (I'm talking about embedded `'` chars. in the literal resulting from PowerShell's own parsing) per se.

Reply:
Given https://github.com/PowerShell/PowerShell/issues/4661, it's probably worth pointing out that whatever PS argument evaluates to the _empty string_ must be represented as `""`.

Reply:
Yes, I was also talking about "embedded `'` chars". It is correct that [these](https://docs.microsoft.com/en-us/cpp/cpp/parsing-cpp-command-line-arguments) rules don't require quoting arguments containing `'`. 

The reason I included `'` in the list of chars that require quoting, is that some applications mostly follow the MSVC rules, but additionally allow `'` as a string delimiter (AFAIK cygwin, ruby). As I see no disadvantage for apps that strictly follow MSVC rules, I thought it might be worth to also quote arguments containing `'`.